### PR TITLE
DCMAW-8456: Create proxy APIgw in lower environments

### PR DESCRIPTION
**Pull request by sandymay-dd**
DCMAW-8456

### What changed
Creates a proxy APIgw in lower environments. This APIgw is a proxy to the Private API under test.

This also includes a proxy lambda to map the X-Custom-Auth header into the Authorization header. Note: this was not possible in APIgw request mapping due to how AWS handles Signature V4

### Why did it change
To enable testing of the Private APIgw outside of a VPC (i.e. by local engineers)

## Checklists

- [x] There is a ticket raised for this PR that is present in the branch name
- [ ] No PII data logged. [See guidance here](https://govukverify.atlassian.net/wiki/spaces/DCMAW/pages/3502407722/PII+Logging+Considerations)
- [x] Demo to a BA, TA, and the team.
- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: backend-api/template.yaml [hunk 1]
```diff
@@ -381,13 +376,178 @@ Resources:
...
+          - REGIONAL
+      RegionalCertificateArn: !Sub '{{resolve:ssm:/${Environment}/Platform/ACM/PrimaryZoneWildcardCertificateARN}}'
+      SecurityPolicy: TLS_1_2
+    Condition: ProxyApiDeployment
+
+  ProxyApiBasePathMapping:
+    Type: AWS::ApiGateway::BasePathMapping
+    DependsOn: ProxyApiDomainName
+    Properties:
+      DomainName: !Sub proxy-${AWS::StackName}.review-b-async.dev.account.gov.uk
```
Comment (sandymay-dd) [inline backend-api/template.yaml:435]: tied to dev - update

File: backend-api/template.yaml [hunk 2]
```diff
@@ -381,13 +376,178 @@ Resources:
...
+      RestApiId: !Ref ProxyApi
+      Stage: !Ref ProxyApi.Stage
+    Condition: ProxyApiDeployment
+
+  ProxyApiRecordSet:
+    Type: AWS::Route53::RecordSet
+    Properties:
+      Name: !Sub proxy-${AWS::StackName}.review-b-async.dev.account.gov.uk
```
Comment (sandymay-dd) [inline backend-api/template.yaml:443]: tied to dev - update

File: backend-api/api-tests/clientCredentialsGrantFlow.test.ts
Comment (sandymay-dd) [inline backend-api/api-tests/clientCredentialsGrantFlow.test.ts:1]: There's a future ticket to add tests for the client credentials grant flow. I've added these to prove that the proxy integration is working

File: backend-api/generate_env_file.sh [hunk 3]
```diff
@@ -11,4 +13,5 @@ echo "Generating .env file for the $SAM_STACK stack"
 
 SELF=$(aws cloudformation describe-stacks --stack-name "$SAM_STACK" --query "Stacks[0].Outputs[?OutputKey=='PublicAPIGWBaseURL'].OutputValue" --output text)
 
-echo "SELF_PUBLIC=$SELF" > .env
\ No newline at end of file
+echo "SELF_PUBLIC=$SELF" > .env
+echo "PROXY_API_URL=https://proxy-$SAM_STACK.review-b-async.$ENV.account.gov.uk" >> .env
```
Comment (jmooney-dd) [inline backend-api/generate_env_file.sh:17]: Just checking we will actually use the prod prefix here? I assume we will as it's a proxy api url for testing

File: backend-api/src/functions/proxy/proxyHandler.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  logger.log("STARTED");
+  if (configResult.isError) {
+    logger.log("ENVIRONMENT_VARIABLE_MISSING", {
+      errorMessage: configResult.value.errorMessage,
+    });
+    return {
+      headers: {
+        "Content-Type": "application/json",
+      },
+      statusCode: 500,
```
Comment (jmooney-dd) [inline backend-api/src/functions/proxy/proxyHandler.ts:34]: As we return 500 more than once is it worth consting the response and then return something like `serverError500Response()`

File: backend-api/generate_env_file.sh (hunk 3 above)
Comment (sandymay-dd) [reply backend-api/generate_env_file.sh:17]: I think your question conflicts with itself!

Comment (sandymay-dd) [reply backend-api/generate_env_file.sh:17]: We will not need this in Prod :)

File: backend-api/src/functions/proxy/proxyHandler.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        error_description: "Server Error",
+      }),
+    };
+  }
+
+  const method = event.httpMethod;
+
+  if (method !== "POST") {
+    logger.log("UNEXPECTED_HTTP_METHOD", {
+      errorMessage: "Method is not an accepted value",
```
Comment (jmooney-dd) [inline backend-api/src/functions/proxy/proxyHandler.ts:65]: If we only accept 'POST' here, is it worth making this message more specific? Super minor thought happy for you to dismiss

File: backend-api/src/functions/proxy/proxyHandler.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    standardisedHeaders["Authorization"] = customAuthHeaderValue;
+  }
+
+  const proxyRequestService = dependencies.proxyRequestService();
+  const proxyRequestResult = await proxyRequestService.makeProxyRequest({
+    backendApiUrl: configResult.value.ASYNC_BACKEND_API_URL,
+    body: event.body,
+    path,
+    headers: standardisedHeaders,
+    method: method,
```
Comment (jmooney-dd) [inline backend-api/src/functions/proxy/proxyHandler.ts:96]:
```suggestion
    method,
```

File: backend-api/src/functions/proxy/proxyHandler.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  if (!apiGwHeaders) return standardisedHeaders;
+  const apiGwHeaderKeys = Object.keys(apiGwHeaders);
+  apiGwHeaderKeys.forEach((headerKey) => {
+    const headerValue = apiGwHeaders[headerKey];
+    if (
+      typeof headerValue === "string" ||
+      typeof headerValue === "number" ||
+      typeof headerValue === "boolean"
+    ) {
+      // This fails TLS validation as the target of the request (private apigw execute-url isn't on the certificate for this DNS)
```
Comment (jmooney-dd) [inline backend-api/src/functions/proxy/proxyHandler.ts:134]: Failing TLS validation sounds bad. Are you saying the next line fixes it? Or is TLS validation going to fail and we're ok with that?

Comment (sandymay-dd) [reply backend-api/src/functions/proxy/proxyHandler.ts:134]: bad wording on my part - will update! Effectively there are a bunch of AWS headers that are included automatically from the Proxy that are included - one causes a definite issue. We should probably try and strip all that we can. I'll update the comment and see whether any more headers can go

File: backend-api/template.yaml [hunk 8]
```diff
@@ -381,13 +386,193 @@ Resources:
...
+    Type: AWS::Lambda::Permission
+    Properties:
+      Action: lambda:InvokeFunction
+      FunctionName: !GetAtt ProxyLambda.Arn
+      Principal: apigateway.amazonaws.com
+      SourceAccount: !Sub ${AWS::AccountId}
+      SourceArn: !Sub arn:aws:execute-api:${AWS::Region}:${AWS::AccountId}:${ProxyApi}/*/*/*
+    Condition: ProxyApiDeployment
+
+  ProxyLambda:
```
Comment (sandymay-dd) [inline backend-api/template.yaml:500]: Concurrency for the lambda

File: backend-api/template.yaml [hunk 9]
```diff
@@ -381,13 +386,193 @@ Resources:
...
+    Type: AWS::Serverless::Function
+    DependsOn:
+      - ProxyLambdaLogGroup
+    Metadata:
+      BuildMethod: esbuild
+      BuildProperties:
+        Minify: true
+        Target: es2022
+        Sourcemap: true
```
Comment (sandymay-dd) [inline backend-api/template.yaml:509]: I think this could be set to false

File: backend-api/template.yaml [hunk 10]
```diff
@@ -381,13 +386,193 @@ Resources:
...
+        EntryPoints:
+          - src/functions/proxy/proxyHandler.ts
+    Properties:
+      Environment:
+        Variables:
+          ASYNC_BACKEND_API_URL: !Sub https://${PrivateApi.RestApiId}.execute-api.eu-west-2.amazonaws.com/${Environment}
```
Comment (sandymay-dd) [inline backend-api/template.yaml:515]: Variable could do with renaming to be something like PRIVATE_BACKEND-API_URL

File: backend-api/template.yaml [hunk 11]
```diff
@@ -381,13 +386,193 @@ Resources:
...
+        - PolicyName: ProxyFunctionLoggingPolicy
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - logs:CreateLogGroup
+                  - logs:CreateLogStream
+                  - logs:PutLogEvents
+                Resource: arn:aws:logs:*:*:*
```
Comment (sandymay-dd) [inline backend-api/template.yaml:559]: Should this be tied to the log group?

File: backend-api/template.yaml (hunk 10 above)
Comment (sandymay-dd) [reply backend-api/template.yaml]: done

File: backend-api/template.yaml [hunk 12]
```diff
@@ -381,13 +419,201 @@ Resources:
...
+  ProxyLambda:
+    Type: AWS::Serverless::Function
+    DependsOn:
+      - ProxyLambdaLogGroup
+    Metadata:
+      BuildMethod: esbuild
+      BuildProperties:
+        Minify: true
+        Target: es2022
+        Sourcemap: true
```
Comment (sandymay-dd) [inline backend-api/template.yaml:550]: change to false

Comment (sandymay-dd) [inline backend-api/template.yaml:550]:
```suggestion
        Sourcemap: false
```

File: backend-api/template.yaml [hunk 13]
```diff
@@ -381,13 +419,201 @@ Resources:
...
+          - src/functions/proxy/proxyHandler.ts
+    Properties:
+      Environment:
+        Variables:
+          PRIVATE_API_URL: !Sub https://${PrivateApi.RestApiId}.execute-api.eu-west-2.amazonaws.com/${Environment}
+      FunctionName: !Sub ${AWS::StackName}-proxy-lambda
+      Runtime: nodejs20.x
+      Handler: proxyHandler.lambdaHandler
+      Role: !GetAtt ProxyLambdaRole.Arn
+      Architectures:
```
Comment (sandymay-dd) [inline backend-api/template.yaml:561]: set globally

File: backend-api/run-tests.sh [hunk 14]
```diff
@@ -12,14 +20,16 @@ else
...
+  else
+    cp -rf results $TEST_REPORT_ABSOLUTE_DIR
+    exit 1
+  fi
+fi
+
+# elif [[ "$TEST_ENVIRONMENT" == "staging" ]]; then
 #   if npm run test:stage --ci --silent; then
 #     cp -rf results $TEST_REPORT_ABSOLUTE_DIR
 #   else
```
Comment (sandymay-dd) [inline backend-api/run-tests.sh:35]:
```suggestion
```

**Review by Pranjal-GDS — APPROVED** [review]

**Pull request merged**